File: backend/src/main/java/com/team05/petmeeting/domain/animal/entity/Animal.java
package com.team05.petmeeting.domain.animal.entity;

import com.team05.petmeeting.domain.animal.dto.external.AnimalItem;
import com.team05.petmeeting.domain.comment.entity.AnimalComment;
import com.team05.petmeeting.domain.shelter.entity.Shelter;
import com.team05.petmeeting.global.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity
@Getter
@Table(name = "animals")
@NoArgsConstructor
@Builder //
@AllArgsConstructor(access = AccessLevel.PRIVATE) // 빌더를 위해 모든 필드 생성자 필요
public class Animal extends BaseEntity {

    @Column(name = "desertion_no", nullable = false, length = 50, unique = true)
    private String desertionNo; // 유기번호

    @Column(name = "process_state", nullable = true, length = 30)
    private String processState; // 상태 (보호중, 입양가능, 입양대기, 파양, 종결 등)

    @Column(name = "state_group", nullable = false)
    private Integer stateGroup; // 0: 보호중, 1: 종료

    @Column(name = "notice_no", nullable = true, length = 50)
    private String noticeNo; // 공고번호

    @Column(name = "notice_edt", nullable = true)
    private LocalDate noticeEdt; // 공고 종료일

    @Column(name = "happen_place", nullable = true, length = 255)
    private String happenPlace; // 발견 장소

    @Column(name = "up_kind_nm", nullable = true, length = 30)
    private String upKindNm; // 종 (개, 고양이 등)

    @Column(name = "kind_full_name", nullable = true, length = 100)
    private String kindFullNm; // 품종 (예: 믹스견, 시바견 등)

    @Column(name = "color_cd", nullable = true, length = 100)
    private String colorCd; // 색상

    @Column(name = "age", nullable = true)
    private String age; // 나이

    @Column(name = "weight", nullable = true, length = 30)
    private String weight; // 몸무게 (예: 5kg, 10kg 등)

    @Column(name = "sex_cd", nullable = true, length = 10)
    private String sexCd; // 성별

    @Column(name = "popfile1", nullable = true, length = 500)
    private String popfile1; // 사진 URL

    @Column(name = "popfile2", nullable = true, length = 500)
    private String popfile2; // 사진 URL

    @Column(name = "special_mark", nullable = true, length = 500)
    private String specialMark; // 사진 URL

    @Column(name = "care_nm", nullable = true)
    private String careNm; // 보호소 이름

    @Column(name = "care_owner_nm", nullable = true, length = 100)
    private String careOwnerNm; // 보호소 담당자

    @Column(name = "care_tel", nullable = true)
    private String careTel; // 보호소 전화번호

    @Column(name = "care_addr", nullable = true)
    private String careAddr; // 보호소 주소

    @Column(name = "total_cheer_count", nullable = false)
    private Integer totalCheerCount; // 응원 수

    @Column(name = "api_updated_at")
    private LocalDateTime apiUpdatedAt;

    @Column(name = "name", nullable = true)
    private String name;

    @OneToMany(mappedBy = "animal", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<AnimalComment> comments = new ArrayList<>();

    @ManyToOne
    private Shelter shelter;

    private Integer determineStateGroup(String processState) {
        if (processState != null && processState.contains("보호")) {
            return 0;
        }
        return 1;
    }

    public void updateName(String name) {
        this.name = name;
    }

    public void updateFrom(AnimalItem item) {
        this.processState = item.getProcessState();
        this.stateGroup = determineStateGroup(item.getProcessState());
        this.noticeNo = item.getNoticeNo();
        this.noticeEdt = parseNoticeEdt(item.getNoticeEdt());
        this.happenPlace = item.getHappenPlace();
        this.upKindNm = item.getUpKindNm();
        this.kindFullNm = item.getKindFullNm();
        this.colorCd = item.getColorCd();
        this.age = item.getAge();
        this.weight = item.getWeight();
        this.sexCd = item.getSexCd();
        this.popfile1 = item.getPopfile1();
        this.popfile2 = item.getPopfile2();
        this.specialMark = item.getSpecialMark();
        this.careNm = item.getCareNm();
        this.careOwnerNm = item.getCareOwnerNm();
        this.careTel = item.getCareTel();
        this.careAddr = item.getCareAddr();
        this.apiUpdatedAt = parseUpdTm(item.getUpdTm());
    }

    public boolean needsUpdateFrom(AnimalItem item) {
        LocalDateTime incomingUpdatedAt = parseUpdTm(item.getUpdTm());

        if (this.apiUpdatedAt != null && incomingUpdatedAt != null) {
            return incomingUpdatedAt.isAfter(this.apiUpdatedAt);
        }

        return !Objects.equals(this.processState, item.getProcessState())
                || !Objects.equals(this.noticeNo, item.getNoticeNo())
                || !Objects.equals(this.happenPlace, item.getHappenPlace())
                || !Objects.equals(this.specialMark, item.getSpecialMark())
                || !Objects.equals(this.careNm, item.getCareNm());
    }

    public void assignShelter(Shelter shelter) {
        this.shelter = shelter;
    }

    private static LocalDateTime parseUpdTm(String updTm) {
        if (updTm == null || updTm.isBlank()) {
            return null;
        }

        return LocalDateTime.parse(
                updTm,
                DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.S")
        );
    }

    private Animal(
            String desertionNo,
            String processState,
            String noticeNo,
            LocalDate noticeEdt,
            String happenPlace,
            String upKindNm,
            String kindFullName,
            String colorCd,
            String age,
            String weight,
            String sexCd,
            String popfile1,
            String popfile2,
            String specialMark,
            String careOwnerNm,
            String careNm,
            String careTel,
            String careAddr,
            Integer totalCheerCount
    ) {
        this.desertionNo = desertionNo;
        this.processState = processState;
        this.stateGroup = determineStateGroup(processState); // 생성 시점에 0 또는 1 할당
        this.noticeNo = noticeNo;
        this.noticeEdt = noticeEdt;
        this.happenPlace = happenPlace;
        this.upKindNm = upKindNm;
        this.kindFullNm = kindFullName;
        this.colorCd = colorCd;
        this.age = age;
        this.weight = weight;
        this.sexCd = sexCd;
        this.popfile1 = popfile1;
        this.popfile2 = popfile2;
        this.specialMark = specialMark;
        this.careOwnerNm = careOwnerNm;
        this.careNm = careNm;
        this.careTel = careTel;
        this.careAddr = careAddr;
        this.totalCheerCount = totalCheerCount;
    }


    public static Animal from(AnimalItem item) {
        Animal animal = new Animal(
                item.getDesertionNo(),
                item.getProcessState(),
                item.getNoticeNo(),
                parseNoticeEdt(item.getNoticeEdt()),
                item.getHappenPlace(),
                item.getUpKindNm(),
                item.getKindFullNm(),
                item.getColorCd(),
                item.getAge(),
                item.getWeight(),
                item.getSexCd(),
                item.getPopfile1(),
                item.getPopfile2(),
                item.getSpecialMark(),
                item.getCareOwnerNm(),
                item.getCareNm(),
                item.getCareTel(),
                item.getCareAddr(),
                0
        );

        animal.apiUpdatedAt = parseUpdTm(item.getUpdTm());
        return animal;
    }

    private static LocalDate parseNoticeEdt(String noticeEdt) {
        if (noticeEdt == null || noticeEdt.isBlank()) {
            return null;
        }

        return LocalDate.parse(noticeEdt, DateTimeFormatter.BASIC_ISO_DATE);
    }

    public double getTemperature() {
        double cheerGoal = 50.0; // todo 통합 후 수정
        if (this.totalCheerCount == null) {
            return 0;
        }
        return this.totalCheerCount / cheerGoal * 100;
    }
}

File: backend/src/main/java/com/team05/petmeeting/domain/animal/entity/Animal.kt
package com.team05.petmeeting.domain.animal.entity

import com.team05.petmeeting.domain.animal.dto.external.AnimalItem
import com.team05.petmeeting.domain.comment.entity.AnimalComment
import com.team05.petmeeting.domain.shelter.entity.Shelter
import com.team05.petmeeting.global.entity.BaseEntity
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Objects

@Entity
@Table(name = "animals")
class Animal() : BaseEntity() {
    @Column(name = "desertion_no", nullable = false, length = 50, unique = true)
    var desertionNo: String? = null
        protected set

    @Column(name = "process_state", nullable = true, length = 30)
    var processState: String? = null
        protected set

    @Column(name = "state_group", nullable = false)
    var stateGroup: Int? = null
        protected set

    @Column(name = "notice_no", nullable = true, length = 50)
    var noticeNo: String? = null
        protected set

    @Column(name = "notice_edt", nullable = true)
    var noticeEdt: LocalDate? = null
        protected set

    @Column(name = "happen_place", nullable = true, length = 255)
    var happenPlace: String? = null
        protected set

    @Column(name = "up_kind_nm", nullable = true, length = 30)
    var upKindNm: String? = null
        protected set

    @Column(name = "kind_full_name", nullable = true, length = 100)
    var kindFullNm: String? = null
        protected set

    @Column(name = "color_cd", nullable = true, length = 100)
    var colorCd: String? = null
        protected set

    @Column(name = "age", nullable = true)
    var age: String? = null
        protected set

    @Column(name = "weight", nullable = true, length = 30)
    var weight: String? = null
        protected set

    @Column(name = "sex_cd", nullable = true, length = 10)
    var sexCd: String? = null
        protected set

    @Column(name = "popfile1", nullable = true, length = 500)
    var popfile1: String? = null
        protected set

    @Column(name = "popfile2", nullable = true, length = 500)
    var popfile2: String? = null
        protected set

    @Column(name = "special_mark", nullable = true, length = 500)
    var specialMark: String? = null
        protected set

    @Column(name = "care_nm", nullable = true)
    var careNm: String? = null
        protected set

    @Column(name = "care_owner_nm", nullable = true, length = 100)
    var careOwnerNm: String? = null
        protected set

    @Column(name = "care_tel", nullable = true)
    var careTel: String? = null
        protected set

    @Column(name = "care_addr", nullable = true)
    var careAddr: String? = null
        protected set

    @Column(name = "total_cheer_count", nullable = false)
    var totalCheerCount: Int = 0
        protected set

    @Column(name = "api_updated_at")
    var apiUpdatedAt: LocalDateTime? = null
        protected set

    @Column(name = "name", nullable = true)
    var name: String? = null
        protected set

    @OneToMany(mappedBy = "animal", cascade = [CascadeType.ALL], orphanRemoval = true)
    var comments: MutableList<AnimalComment> = mutableListOf()
        protected set

    @ManyToOne
    var shelter: Shelter? = null
        protected set

    private constructor(builder: Builder) : this() {
        desertionNo = builder.desertionNo
        processState = builder.processState
        stateGroup = builder.stateGroup ?: determineStateGroup(builder.processState)
        noticeNo = builder.noticeNo
        noticeEdt = builder.noticeEdt
        happenPlace = builder.happenPlace
        upKindNm = builder.upKindNm
        kindFullNm = builder.kindFullNm
        colorCd = builder.colorCd
        age = builder.age
        weight = builder.weight
        sexCd = builder.sexCd
        popfile1 = builder.popfile1
        popfile2 = builder.popfile2
        specialMark = builder.specialMark
        careOwnerNm = builder.careOwnerNm
        careNm = builder.careNm
        careTel = builder.careTel
        careAddr = builder.careAddr
        totalCheerCount = builder.totalCheerCount ?: 0
        apiUpdatedAt = builder.apiUpdatedAt
        name = builder.name
        comments = builder.comments?.toMutableList() ?: mutableListOf()
        shelter = builder.shelter
    }

    private fun determineStateGroup(processState: String?): Int =
        if (processState != null && processState.contains("보호")) 0 else 1

    fun updateName(name: String?) {
        this.name = name
    }

    fun updateFrom(item: AnimalItem) {
        processState = item.processState
        stateGroup = determineStateGroup(item.processState)
        noticeNo = item.noticeNo
        noticeEdt = parseNoticeEdt(item.noticeEdt)
        happenPlace = item.happenPlace
        upKindNm = item.upKindNm
        kindFullNm = item.kindFullNm
        colorCd = item.colorCd
        age = item.age
        weight = item.weight
        sexCd = item.sexCd
        popfile1 = item.popfile1
        popfile2 = item.popfile2
        specialMark = item.specialMark
        careNm = item.careNm
        careOwnerNm = item.careOwnerNm
        careTel = item.careTel
        careAddr = item.careAddr
        apiUpdatedAt = parseUpdTm(item.updTm)
    }

    fun needsUpdateFrom(item: AnimalItem): Boolean {
        val incomingUpdatedAt = parseUpdTm(item.updTm)

        if (apiUpdatedAt != null && incomingUpdatedAt != null) {
            return incomingUpdatedAt.isAfter(apiUpdatedAt)
        }

        return !Objects.equals(processState, item.processState) ||
            !Objects.equals(noticeNo, item.noticeNo) ||
            !Objects.equals(happenPlace, item.happenPlace) ||
            !Objects.equals(specialMark, item.specialMark) ||
            !Objects.equals(careNm, item.careNm)
    }

    fun assignShelter(shelter: Shelter) {
        this.shelter = shelter
    }

    fun getTemperature(): Double {
        val cheerGoal = 50.0
        return totalCheerCount / cheerGoal * 100
    }

    class Builder {
        internal var desertionNo: String? = null
        internal var processState: String? = null
        internal var stateGroup: Int? = null
        internal var noticeNo: String? = null
        internal var noticeEdt: LocalDate? = null
        internal var happenPlace: String? = null
        internal var upKindNm: String? = null
        internal var kindFullNm: String? = null
        internal var colorCd: String? = null
        internal var age: String? = null
        internal var weight: String? = null
        internal var sexCd: String? = null
        internal var popfile1: String? = null
        internal var popfile2: String? = null
        internal var specialMark: String? = null
        internal var careOwnerNm: String? = null
        internal var careNm: String? = null
        internal var careTel: String? = null
        internal var careAddr: String? = null
        internal var totalCheerCount: Int? = null
        internal var apiUpdatedAt: LocalDateTime? = null
        internal var name: String? = null
        internal var comments: List<AnimalComment>? = null
        internal var shelter: Shelter? = null

        fun desertionNo(desertionNo: String?) = apply { this.desertionNo = desertionNo }
        fun processState(processState: String?) = apply { this.processState = processState }
        fun stateGroup(stateGroup: Int?) = apply { this.stateGroup = stateGroup }
        fun noticeNo(noticeNo: String?) = apply { this.noticeNo = noticeNo }
        fun noticeEdt(noticeEdt: LocalDate?) = apply { this.noticeEdt = noticeEdt }
        fun happenPlace(happenPlace: String?) = apply { this.happenPlace = happenPlace }
        fun upKindNm(upKindNm: String?) = apply { this.upKindNm = upKindNm }
        fun kindFullNm(kindFullNm: String?) = apply { this.kindFullNm = kindFullNm }
        fun colorCd(colorCd: String?) = apply { this.colorCd = colorCd }
        fun age(age: String?) = apply { this.age = age }
        fun weight(weight: String?) = apply { this.weight = weight }
        fun sexCd(sexCd: String?) = apply { this.sexCd = sexCd }
        fun popfile1(popfile1: String?) = apply { this.popfile1 = popfile1 }
        fun popfile2(popfile2: String?) = apply { this.popfile2 = popfile2 }
        fun specialMark(specialMark: String?) = apply { this.specialMark = specialMark }
        fun careOwnerNm(careOwnerNm: String?) = apply { this.careOwnerNm = careOwnerNm }
        fun careNm(careNm: String?) = apply { this.careNm = careNm }
        fun careTel(careTel: String?) = apply { this.careTel = careTel }
        fun careAddr(careAddr: String?) = apply { this.careAddr = careAddr }
        fun totalCheerCount(totalCheerCount: Int?) = apply { this.totalCheerCount = totalCheerCount }
        fun apiUpdatedAt(apiUpdatedAt: LocalDateTime?) = apply { this.apiUpdatedAt = apiUpdatedAt }
        fun name(name: String?) = apply { this.name = name }
        fun comments(comments: List<AnimalComment>?) = apply { this.comments = comments }
        fun shelter(shelter: Shelter?) = apply { this.shelter = shelter }
        fun build(): Animal = Animal(this)
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()

        @JvmStatic
        fun from(item: AnimalItem): Animal {
            val animal = Builder()
                .desertionNo(item.desertionNo)
                .processState(item.processState)
                .noticeNo(item.noticeNo)
                .noticeEdt(parseNoticeEdt(item.noticeEdt))
                .happenPlace(item.happenPlace)
                .upKindNm(item.upKindNm)
                .kindFullNm(item.kindFullNm)
                .colorCd(item.colorCd)
                .age(item.age)
                .weight(item.weight)
                .sexCd(item.sexCd)
                .popfile1(item.popfile1)
                .popfile2(item.popfile2)
                .specialMark(item.specialMark)
                .careOwnerNm(item.careOwnerNm)
                .careNm(item.careNm)
                .careTel(item.careTel)
                .careAddr(item.careAddr)
                .totalCheerCount(0)
                .build()

            animal.apiUpdatedAt = parseUpdTm(item.updTm)
            return animal
        }

        private fun parseUpdTm(updTm: String?): LocalDateTime? {
            if (updTm.isNullOrBlank()) {
                return null
            }

            return LocalDateTime.parse(
                updTm,
                DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.S"),
            )
        }

        private fun parseNoticeEdt(noticeEdt: String?): LocalDate? {
            if (noticeEdt.isNullOrBlank()) {
                return null
            }

            return LocalDate.parse(noticeEdt, DateTimeFormatter.BASIC_ISO_DATE)
        }
    }
}
